### ISSUE TITLE
Allow attributes to be passed to drupal_add_[css|js] (SRI)

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5478]
This is the same as https://www.drupal.org/project/drupal/issues/1664602

> ## Problem/Motivation
> Currently there is no way to pass `#attributes` to the `drupal_add_js()` function. This is not a feature that is widely used, however, there are unique cases where the `<script>` tag generated needs to have available an `id=""` or other unique attributes to that tag.
> 
> ## Proposed resolution
> By initializing the `#attributes` key of the `$js_element` array we can easily pass specific `#attribute` elements in our `drupal_add_js()` calls, and since `theme('html_tag', array('element' => $js_element));` handles `#attributes` passed to it, all additional support is inherited. Here is an example of the new attribute code working in the case 'external' type of `drupal_add_js()` is invoked. The patch of course accounts for all cases of `'type'`:

```php
case 'external':
  $js_element = $element;

  // Attach custom attributes to the element.
  $js_element['#attributes'] = !empty($item['attributes']) ? $item['attributes'] : array();

  // Preprocessing for external JavaScript files is ignored.
  if ($item['defer']) {
    $js_element['#attributes']['defer'] = 'defer';
  }

  $js_element['#attributes']['src'] = $item['data'];
  $processed[$index++] = theme('html_tag', array('element' => $js_element));
  break;
```

**Comment by klonos** [comment]
In the D7 issue there are links to multiple contrib issues, which show that popular modules like reCAPTCHA and themes like Omega would benefit by this improvement.

**Comment by indigoxela** [comment]
@klonos attributes have been added to backdrop_add_js in #2877 and to backdrop_add_css in #4981

Is there anything left to do here?

**Comment by klonos** [comment]
You are right @indigoxela ...had I searched our issue queue for `1664602`, then I would have seen #2877 and #4981.

> Is there anything left to do here?

I guess that we'll need to check https://www.drupal.org/project/drupal/issues/1664602 for anything that we may have missed in our implementation. For example, the last comment there (5 months ago at the time of writing this) suggests that "doesn't seem to support preprocessing of JS files.". I haven't confirmed that yet, but my point is that we should check for things reported/discussed there.

**Comment by klonos** [comment]
OK, I've compared the D7 patch against the code changes that went into Backdrop core with:
- https://github.com/backdrop/backdrop/commit/9058d7cec7ea49dd54c5d6482347e1c9105d95b7
- https://github.com/backdrop/backdrop/commit/927b188a5fef07095bdea423d512d043880dc213

So most of the work for this is already done in Backdrop, but I have found the following differences:
- In `backdrop_css_defaults()` we've added `'attributes'` as an empty array, whereas in the Drupal patch they are adding the array with `'type' => 'text/css'`. Since in HTML5 [specifying the type in `<link>` tags is optional](https://stackoverflow.com/questions/5409114/is-type-text-css-necessary-in-a-link-tag), we're OK 👍🏼 
- (similarly) In `backdrop_js_defaults()` we've added `'attributes'` as an empty array, whereas in the Drupal patch they are adding the array with `'type' => 'text/javascript'`. Since in HTML5 [specifying the type in `<script>` tags is optional](https://stackoverflow.com/questions/4195427/is-the-type-attribute-necessary-for-script-tags), we're OK 👍🏼 
- In the D7 patch, they are adding some defensive coding in various places, to "Ensure the item's attributes setting is an array before attempting to merge with the element's attributes.".
- In `backdrop_add_js()` we are "manually" adding certain keys to arrays where we could be adding them via `backdrop_js_defaults()` (and then only add those keys that differ from the defaults).
- For `backdrop_add_css()`, we've added the same simplistic set of tests as D7, testing things with CSS aggregation turned off, which is OK, but I would like us to consider expanding the tests to the same level as the tests added for `backdrop_add_js()`.

PR for consideration: https://github.com/backdrop/backdrop/pull/3945

PS: there's another PR where I've also cleaned up inline comments and docblocks around the relevant functions: https://github.com/backdrop/backdrop/pull/3944

**Comment by yorkshire-pudding** [comment]
@klonos - Doesn't #4981 cover the `backdrop_add_css` element?

**Comment by klonos** [comment]
@yorkshire-pudding yes. Please see the same question asked by @indigoxela back in https://github.com/backdrop/backdrop-issues/issues/5478#issuecomment-1024890224 and my reply in the comment right after that.

Although I haven't touched this issue in over a year, from my comments in the comment right before yours, and from looking at the linked PR, my understanding is that this is an effort to follow up on the things that we've added into Backdrop core with #2877 and #4981, and either improve/optimize or add anything that we might have missed that can be useful.

I have rebased the PR, so this still needs code review and potentially some testing (to make sure that we don't break things).

All tests are green, besides a coding standards fail, which is about an array exceeding max length - but that's in an unrelated portion of the code, in function `backdrop_common_theme()`, which includes many such lines. So if we are to fix, it should be in a follow-up issue, to address all array declarations that go beyond 80 characters - not only the line pointed out by PHPCS in this PR.
